## Final Project Submission

Please fill out:
* Student name:                           ``Dahir Ahmed``
* Student pace:                             ``Part time`` 
* Scheduled project review date/time:       ``27TH OCT TO  31ST OCT 2025``
* Instructor name:                           ``Brian Chacha``
* Blog post URL:


In [14]:
# Your code here - remember to use markdown cells for comments as well!
# Import the pandas library for data manipulation and analysis
import pandas as pd

# Import the numpy library for numerical operations and working with arrays
import numpy as np


# Import the warnings module to control or suppress warning messages
import warnings

# Suppress all warning messages to keep the output clean
warnings.filterwarnings('ignore')



1.1) Read Aviation_Data.csv into a pandas DataFrame named df

In [15]:
# Load the CSV file "Aviation_Data.csv" from the 'data' folder into a pandas DataFrame named 'df'

df = pd.read_csv("./data/Aviation_Data.csv")

# Display the first 5 rows of the DataFrame to quickly inspect the data

df.head()

Event.Id Investigation.Type Accident.Number  Event.Date  \
0  20001218X45444           Accident      SEA87LA080  1948-10-24   
1  20001218X45447           Accident      LAX94LA336  1962-07-19   
2  20061025X01555           Accident      NYC07LA005  1974-08-30   
3  20001218X45448           Accident      LAX96LA321  1977-06-19   
4  20041105X01764           Accident      CHI79FA064  1979-08-02   

          Location        Country Latitude Longitude Airport.Code  \
0  MOOSE CREEK, ID  United States      NaN       NaN          NaN   
1   BRIDGEPORT, CA  United States      NaN       NaN          NaN   
2    Saltville, VA  United States  36.9222  -81.8781          NaN   
3       EUREKA, CA  United States      NaN       NaN          NaN   
4       Canton, OH  United States      NaN       NaN          NaN   

  Airport.Name  ... Purpose.of.flight Air.carrier Total.Fatal.Injuries  \
0          NaN  ...          Personal         NaN                  2.0   
1          NaN  ...          Personal         NaN                  4.0   
2          NaN  ...          Personal         NaN                  3.0   
3          NaN  ...          Personal         NaN                  2.0   
4          NaN  ...          Personal         NaN                  1.0   

  Total.Serious.Injuries Total.Minor.Injuries Total.Uninjured  \
0                    0.0                  0.0             0.0   
1                    0.0                  0.0             0.0   
2                    NaN                  NaN             NaN   
3                    0.0                  0.0             0.0   
4                    2.0                  NaN             0.0   

  Weather.Condition  Broad.phase.of.flight   Report.Status Publication.Date  
0               UNK                 Cruise  Probable Cause              NaN  
1               UNK                Unknown  Probable Cause       19-09-1996  
2               IMC                 Cruise  Probable Cause       26-02-2007  
3               IMC                 Cruise  Probable Cause       12-09-2000  
4               VMC               Approach  Probable Cause       16-04-1980  

[5 rows x 31 columns]

In [16]:
# Display the shape (rows, columns)
df.shape

# Print the number of rows and columns in a readable format
print(f"This dataset has {df.shape[0]} rows and {df.shape[1]} columns.")


This dataset has 90348 rows and 31 columns.


In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 90348 entries, 0 to 90347
Data columns (total 31 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Event.Id                88889 non-null  object 
 1   Investigation.Type      90348 non-null  object 
 2   Accident.Number         88889 non-null  object 
 3   Event.Date              88889 non-null  object 
 4   Location                88837 non-null  object 
 5   Country                 88663 non-null  object 
 6   Latitude                34382 non-null  object 
 7   Longitude               34373 non-null  object 
 8   Airport.Code            50249 non-null  object 
 9   Airport.Name            52790 non-null  object 
 10  Injury.Severity         87889 non-null  object 
 11  Aircraft.damage         85695 non-null  object 
 12  Aircraft.Category       32287 non-null  object 
 13  Registration.Number     87572 non-null  object 
 14  Make                    88826 non-null

# How to Find Duplicates in our Aviation Data 

our data has **90,348 rows** and **31 columns**.  
Sometimes the **same crash** gets entered **twice** by mistake.  
We want to **find & remove** those duplicates so our analysis is clean!



In [20]:
duplicate_rows = df.duplicated().sum()

print(f"There are {duplicate_rows} duplicate rows in this dataset.")


There are 1390 duplicate rows in this dataset.


In [21]:
#If you don’t want to modify the original DataFrame:

cleaned_df = df.drop_duplicates()

In [23]:
cleaned_df.shape

(88958, 31)

## How to Find Missing Values in Our Aviation Data

Our dataset(cleaned) has ``88,958`` rows and ``31`` columns.
Sometimes, during data entry, some information gets left blank — like the aircraft type, location, or number of fatalities.

These missing values ``(NaNs)`` can affect our analysis, so we need to identify and handle them before moving forward!

## We’ll check:

1. How many missing values exist in each column

2. Which columns have the most missing data

3. And decide whether to fill, drop, or keep them depending on their importance.

In [24]:
#Check missing data percentages:
cleaned_df.isnull().mean().sort_values(ascending=False) * 100 


Schedule                  85.856247
Air.carrier               81.285550
FAR.Description           64.002113
Aircraft.Category         63.705344
Longitude                 61.360417
Latitude                  61.350300
Airport.Code              43.513793
Airport.Name              40.657389
Broad.phase.of.flight     30.614447
Publication.Date          17.198004
Total.Serious.Injuries    14.140381
Total.Minor.Injuries      13.491760
Total.Fatal.Injuries      12.893725
Engine.Type                8.033004
Report.Status              7.250613
Purpose.of.flight          7.038153
Number.of.Engines          6.916747
Total.Uninjured            6.723398
Weather.Condition          5.127139
Aircraft.damage            3.668023
Registration.Number        1.558039
Injury.Severity            1.201691
Country                    0.331617
Amateur.Built              0.192226
Model                      0.180984
Make                       0.148385
Location                   0.136019
Event.Date                 0

In [29]:
cleaned_df.describe()


Number.of.Engines  Total.Fatal.Injuries  Total.Serious.Injuries  \
count       82805.000000          77488.000000            76379.000000   
mean            1.146585              0.647855                0.279881   
std             0.446510              5.485960                1.544084   
min             0.000000              0.000000                0.000000   
25%             1.000000              0.000000                0.000000   
50%             1.000000              0.000000                0.000000   
75%             1.000000              0.000000                0.000000   
max             8.000000            349.000000              161.000000   

       Total.Minor.Injuries  Total.Uninjured  
count          76956.000000     82977.000000  
mean               0.357061         5.325440  
std                2.235625        27.913634  
min                0.000000         0.000000  
25%                0.000000         0.000000  
50%                0.000000         1.000000  
75%                0.000000         2.000000  
max              380.000000       699.000000

In [30]:
#Get summary stats
cleaned_df.describe(include='all')

Event.Id Investigation.Type Accident.Number  Event.Date  \
count            88889              88958           88889       88889   
unique           87951                 71           88863       14782   
top     20001212X19172           Accident      WPR22FA309  1982-05-16   
freq                 3              85015               2          25   
mean               NaN                NaN             NaN         NaN   
std                NaN                NaN             NaN         NaN   
min                NaN                NaN             NaN         NaN   
25%                NaN                NaN             NaN         NaN   
50%                NaN                NaN             NaN         NaN   
75%                NaN                NaN             NaN         NaN   
max                NaN                NaN             NaN         NaN   

             Location        Country Latitude Longitude Airport.Code  \
count           88837          88663    34382     34373        50249   
unique          27758            219    25592     27156        10375   
top     ANCHORAGE, AK  United States  332739N  0112457W         NONE   
freq              434          82248       19        24         1488   
mean              NaN            NaN      NaN       NaN          NaN   
std               NaN            NaN      NaN       NaN          NaN   
min               NaN            NaN      NaN       NaN          NaN   
25%               NaN            NaN      NaN       NaN          NaN   
50%               NaN            NaN      NaN       NaN          NaN   
75%               NaN            NaN      NaN       NaN          NaN   
max               NaN            NaN      NaN       NaN          NaN   

       Airport.Name  ... Purpose.of.flight Air.carrier Total.Fatal.Injuries  \
count         52790  ...             82697       16648         77488.000000   
unique        24871  ...                26       13590                  NaN   
top         Private  ...          Personal       Pilot                  NaN   
freq            240  ...             49448         258                  NaN   
mean            NaN  ...               NaN         NaN             0.647855   
std             NaN  ...               NaN         NaN             5.485960   
min             NaN  ...               NaN         NaN             0.000000   
25%             NaN  ...               NaN         NaN             0.000000   
50%             NaN  ...               NaN         NaN             0.000000   
75%             NaN  ...               NaN         NaN             0.000000   
max             NaN  ...               NaN         NaN           349.000000   

       Total.Serious.Injuries Total.Minor.Injuries Total.Uninjured  \
count            76379.000000         76956.000000    82977.000000   
unique                    NaN                  NaN             NaN   
top                       NaN                  NaN             NaN   
freq                      NaN                  NaN             NaN   
mean                 0.279881             0.357061        5.325440   
std                  1.544084             2.235625       27.913634   
min                  0.000000             0.000000        0.000000   
25%                  0.000000             0.000000        0.000000   
50%                  0.000000             0.000000        1.000000   
75%                  0.000000             0.000000        2.000000   
max                161.000000           380.000000      699.000000   

       Weather.Condition  Broad.phase.of.flight   Report.Status  \
count              84397                  61724           82508   
unique                 4                     12           17007   
top                  VMC                Landing  Probable Cause   
freq               77303                  15428           61754   
mean                 NaN                    NaN             NaN   
std                  NaN                    NaN             NaN   
min          

### 1. Columns to **DROP** (too many blanks + not very helpful)

| Column                  | % Missing | Why drop it? |
|-------------------------|-----------|--------------|
| **Schedule**            | 86%       | Almost all rows are blank. It just says if the flight was on a regular schedule (like a bus). We already know from `Purpose.of.flight` if it was a normal passenger trip. |
| **Air.carrier**         | 82%       | Mostly empty. It's just the airline name (like "Delta"). We don’t need the name to understand the crash. |
| **FAR.Description**     | 65%       | A boring legal rule number. Missing for most small planes. Doesn’t help us predict injuries or damage. |
| **Aircraft.Category**   | 64%       | Says if it's a plane, helicopter, etc. But we already have `Make` and `Model` — that’s enough! |
| **Longitude / Latitude**| ~62%      | GPS numbers. Great for Google Maps, but more than half are missing. We still have city and airport to know *where*. |
| **Airport.Code**        | 44%       | The short code like "LAX" or "JFK". Many crashes happen **away** from airports, so it’s often blank. |
| **Airport.Name**        | 42%       | Same as above. We can just use the city name instead. |
| **Broad.phase.of.flight**| 32%    | Tells you *when* in the flight it crashed (takeoff? landing?). Helpful, but too many blanks and too many options. Hard to guess the missing ones. |
| **Publication.Date**    | 18%       | This is the day the **report** came out — not the crash day. Totally useless for understanding the accident. |

---



In [33]:
cols_to_drop = [
    'Schedule', 'Air.carrier', 'FAR.Description', 'Aircraft.Category',
    'Longitude', 'Latitude', 'Airport.Code', 'Airport.Name',
    'Broad.phase.of.flight', 'Publication.Date'
]

cleaned_df= cleaned_df.drop(columns=cols_to_drop)

In [36]:
cleaned_df.isnull().mean().sort_values(ascending=False) * 100 

Total.Serious.Injuries    14.140381
Total.Minor.Injuries      13.491760
Total.Fatal.Injuries      12.893725
Engine.Type                8.033004
Report.Status              7.250613
Purpose.of.flight          7.038153
Number.of.Engines          6.916747
Total.Uninjured            6.723398
Weather.Condition          5.127139
Aircraft.damage            3.668023
Registration.Number        1.558039
Injury.Severity            1.201691
Country                    0.331617
Amateur.Built              0.192226
Model                      0.180984
Make                       0.148385
Location                   0.136019
Event.Date                 0.077565
Accident.Number            0.077565
Event.Id                   0.077565
Investigation.Type         0.000000
dtype: float64

### 2. Columns to **KEEP** (a little blank is okay — they’re super important!)

| Column | % Missing | Why keep it? |
|--------|-----------|--------------|
| **Total.Serious.Injuries** | 15% | This tells us how many people got **really hurt**. Most crashes have 0 — so we can just fill blanks with **0**. |
| **Total.Minor.Injuries** | 15% | Same idea — small cuts or bruises. Fill blanks with **0**. |
| **Total.Fatal.Injuries** | 14% | **Super important!** How many people died. We *must* keep this. Fill blanks with **0**. |
| **Engine.Type** | 9% | Is it a normal engine or a jet? Big difference in crashes. Easy to fill with “Unknown” if missing. |
| **Report.Status** | 9% | Says if the report is “final” or “still being checked”. Helps us know if the data is ready. |
| **Purpose.of.flight** | 8% | Was it a normal passenger flight? Training? Crop dusting? **Tells us a lot** about the risk. |
| **Number.of.Engines** | 8% | 1 engine or 2+? Two engines = safer usually. Easy to guess if missing. |
| **Total.Uninjured** | 8% | How many people walked away fine. Completes the story! Fill with a number if missing. |
| **Weather.Condition** | 7% | Was it sunny or stormy? Weather causes many crashes. Most are “good weather” — so fill with that. |
| **Aircraft.damage** | 5% | Did the plane get scratched or totally destroyed? **Main thing** we want to predict. |
| **Registration.Number** | 3% | The plane’s license plate (like N123AB). Helps find duplicates. |
| **Injury.Severity** | 3% | Says “Fatal” or “Minor” in one word. We can make it from the injury counts if needed. |
| **Country, Amateur.Built, Model, Make, Location, Event.Date, Accident.Number, Event.Id** | ≤ 2% | Almost no blanks! These are **must-haves**: 



In [55]:
# checking country which appears most 
cleaned_df['Country'].value_counts().head()

United States     82543
Brazil              374
Canada              359
Mexico              358
United Kingdom      344
Name: Country, dtype: int64

In [54]:
#any missing value will be replaced with mode country which is United states
cleaned_df['Country'].fillna('United States', inplace=True)


In [56]:
cleaned_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 88958 entries, 0 to 90347
Data columns (total 21 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Event.Id                88889 non-null  object 
 1   Investigation.Type      88958 non-null  object 
 2   Accident.Number         88889 non-null  object 
 3   Event.Date              88889 non-null  object 
 4   Location                88837 non-null  object 
 5   Country                 88958 non-null  object 
 6   Injury.Severity         87889 non-null  object 
 7   Aircraft.damage         85695 non-null  object 
 8   Registration.Number     87572 non-null  object 
 9   Make                    88826 non-null  object 
 10  Model                   88797 non-null  object 
 11  Amateur.Built           88787 non-null  object 
 12  Number.of.Engines       82805 non-null  float64
 13  Engine.Type             81812 non-null  object 
 14  Purpose.of.flight       82697 non-null

In [91]:
# lets get numeric column to see use mean mode or median to fill empty cells
numeric_cols = cleaned_df.select_dtypes(include=['number']).columns
print(numeric_cols.tolist())


['Number.of.Engines', 'Total.Fatal.Injuries', 'Total.Serious.Injuries', 'Total.Minor.Injuries', 'Total.Uninjured']


In [87]:
# Will use the number of engine using make
mode_value = cleaned_df['Number.of.Engines'].mode()[0]
median_value = cleaned_df['Number.of.Engines'].median()
mean_value = cleaned_df['Number.of.Engines'].mean()

print(f"Mode: {mode_value}, Median: {median_value}, Mean: {mean_value:.2f}")


Mode: 1.0, Median: 1.0, Mean: 1.15


In [88]:
# Replace missing or zero engine counts with 1
cleaned_df['Number.of.Engines'] = cleaned_df['Number.of.Engines'].replace(0, np.nan)
cleaned_df['Number.of.Engines'].fillna(1, inplace=True)

## Fill missing injury counts with zero

In [ ]:
injury_cols = ['Total.Fatal.Injuries', 
               'Total.Serious.Injuries', 
               'Total.Minor.Injuries', 
               'Total.Uninjured']

# Replace NaN with 0
cleaned_df[injury_cols] = cleaned_df[injury_cols].fillna(0)


In [93]:
cleaned_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 88958 entries, 0 to 90347
Data columns (total 21 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Event.Id                88889 non-null  object 
 1   Investigation.Type      88958 non-null  object 
 2   Accident.Number         88889 non-null  object 
 3   Event.Date              88889 non-null  object 
 4   Location                88837 non-null  object 
 5   Country                 88958 non-null  object 
 6   Injury.Severity         87889 non-null  object 
 7   Aircraft.damage         85695 non-null  object 
 8   Registration.Number     87572 non-null  object 
 9   Make                    88826 non-null  object 
 10  Model                   88797 non-null  object 
 11  Amateur.Built           88787 non-null  object 
 12  Number.of.Engines       88958 non-null  float64
 13  Engine.Type             81812 non-null  object 
 14  Purpose.of.flight       82697 non-null

`` our Investigation.Type column and number has many entries of 88958 but generally the event id has 88889 ``
 

In [97]:
cleaned_df['Investigation.Type'].value_counts()

Accident      85015
Incident       3874
20-09-2021        1
08-12-2022        1
11-12-2020        1
              ...  
18-11-2020        1
13-09-2021        1
02-12-2022        1
01-11-2022        1
20-05-2021        1
Name: Investigation.Type, Length: 71, dtype: int64

#### cleaning event id where is null will solve the problem
`` our Investigation.Type column and number has many entries of 88958 but generally the event id has 88889 ``
 

In [ ]:
# cleaning event id where is null will solve the problem
cleaned_df['Event.Id'].isna().sum()

69

In [101]:
# Drop rows where Event.Id is missing
cleaned_df = cleaned_df.dropna(subset=['Event.Id']).reset_index(drop=True)

In [102]:
cleaned_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 88889 entries, 0 to 88888
Data columns (total 21 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Event.Id                88889 non-null  object 
 1   Investigation.Type      88889 non-null  object 
 2   Accident.Number         88889 non-null  object 
 3   Event.Date              88889 non-null  object 
 4   Location                88837 non-null  object 
 5   Country                 88889 non-null  object 
 6   Injury.Severity         87889 non-null  object 
 7   Aircraft.damage         85695 non-null  object 
 8   Registration.Number     87572 non-null  object 
 9   Make                    88826 non-null  object 
 10  Model                   88797 non-null  object 
 11  Amateur.Built           88787 non-null  object 
 12  Number.of.Engines       88889 non-null  float64
 13  Engine.Type             81812 non-null  object 
 14  Purpose.of.flight       82697 non-null

In [105]:
cleaned_df['Report.Status'].value_counts()

Probable Cause                                                                                                                                                                                                                                                 61754
Foreign                                                                                                                                                                                                                                                         1999
<br /><br />                                                                                                                                                                                                                                                     167
Factual                                                                                                                                                                                                                  

In [106]:
cleaned_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 88889 entries, 0 to 88888
Data columns (total 21 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Event.Id                88889 non-null  object 
 1   Investigation.Type      88889 non-null  object 
 2   Accident.Number         88889 non-null  object 
 3   Event.Date              88889 non-null  object 
 4   Location                88837 non-null  object 
 5   Country                 88889 non-null  object 
 6   Injury.Severity         87889 non-null  object 
 7   Aircraft.damage         85695 non-null  object 
 8   Registration.Number     87572 non-null  object 
 9   Make                    88826 non-null  object 
 10  Model                   88797 non-null  object 
 11  Amateur.Built           88787 non-null  object 
 12  Number.of.Engines       88889 non-null  float64
 13  Engine.Type             81812 non-null  object 
 14  Purpose.of.flight       82697 non-null

In [ ]:
cleaned_df['Purpose.of.flight'].value_counts(normalize=True)*100

Personal                     59.794188
Instructional                12.819087
Unknown                       8.225208
Aerial Application            5.697909
Business                      4.858701
Positioning                   1.990399
Other Work Use                1.528471
Ferry                         0.981898
Aerial Observation            0.960132
Public Aircraft               0.870648
Executive/corporate           0.668706
Flight Test                   0.489740
Skydiving                     0.220081
External Load                 0.148736
Public Aircraft - Federal     0.126970
Banner Tow                    0.122133
Air Race show                 0.119714
Public Aircraft - Local       0.089483
Public Aircraft - State       0.077391
Air Race/show                 0.071345
Glider Tow                    0.064089
Firefighting                  0.048369
Air Drop                      0.013302
ASHO                          0.007255
PUBS                          0.004837
PUBL                     

In [ ]:
cleaned_df['Aircraft.damage'].value_counts(normalize=True)*100

1982-05-16    0.028125
2000-07-08    0.028125
1984-06-30    0.028125
1983-06-05    0.027000
1986-05-17    0.027000
                ...   
2001-12-23    0.001125
2021-12-20    0.001125
2019-11-05    0.001125
2020-05-05    0.001125
2016-02-29    0.001125
Name: Event.Date, Length: 14782, dtype: float64

In [113]:
cleaned_df['Event.Date'] = pd.to_datetime(cleaned_df['Event.Date'], errors='coerce')

# Extract year
cleaned_df['Year'] = cleaned_df['Event.Date'].dt.year

# Show year distribution as percentages
year_counts = cleaned_df['Year'].value_counts(normalize=True) * 100
print(year_counts)


1982    4.042120
1983    4.000495
1984    3.889120
1985    3.482996
1986    3.239996
1987    3.181496
1988    3.071246
1989    2.861996
1990    2.832746
1991    2.769747
1992    2.649372
1993    2.602122
1995    2.597622
1994    2.539122
1998    2.504247
2000    2.497497
1999    2.485122
1996    2.460372
1997    2.416497
2003    2.345622
2001    2.320872
2005    2.284872
2002    2.272497
2007    2.267997
2004    2.195997
2008    2.129622
2006    2.082372
2011    2.081247
2012    2.064372
2010    2.009247
2009    2.005872
2018    1.891123
2016    1.871998
2017    1.842748
2019    1.826998
2022    1.807873
2015    1.779748
2013    1.756123
2021    1.738123
2014    1.726873
2020    1.565998
1979    0.002250
1977    0.001125
1948    0.001125
1981    0.001125
1962    0.001125
1974    0.001125
Name: Year, dtype: float64
